lib/Async/Registry/promise.h

```diff
@@ -76,22 +87,61 @@ auto inspect(Inspector& f, State& x) {
                                  State::Deleted, "Deleted");
 }
 
+using AsyncWaiter = void*;
+using SyncWaiter = std::thread::id;
+struct AsyncWaiterTmp {
+  AsyncWaiter item;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, AsyncWaiterTmp& x) {
+  return f.object(x).fields(
+      f.field("async", reinterpret_cast<intptr_t>(x.item)));
+}
+struct SyncWaiterTmp {
+  SyncWaiter item;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, SyncWaiterTmp& x) {
+  return f.object(x).fields(f.field("sync", fmt::format("{}", x.item)));
+}
+struct WaiterTmp : std::variant<AsyncWaiterTmp, SyncWaiterTmp> {};
+template<typename Inspector>
+auto inspect(Inspector& f, WaiterTmp& x) {
+  return f.variant(x).unqualified().alternatives(
+      inspection::inlineType<AsyncWaiterTmp>(),
+      inspection::inlineType<SyncWaiterTmp>());
+}
+struct Waiter : std::variant<AsyncWaiter, SyncWaiter> {};
+template<typename Inspector>
+auto inspect(Inspector& f, Waiter& x) {
+  if constexpr (!Inspector::isLoading) {  // only serialize
+    WaiterTmp tmp = std::visit(
+        overloaded{
+            [&](AsyncWaiter waiter) {
+              return WaiterTmp{AsyncWaiterTmp{waiter}};
+            },
+            [&](SyncWaiter waiter) { return WaiterTmp{SyncWaiterTmp{waiter}}; },
+        },
+        x);
+    return f.apply(tmp);
+  }
+}
+
 struct PromiseSnapshot {
   void* id;
   Thread thread;
   SourceLocationSnapshot source_location;
-  void* waiter;
+  std::optional<Waiter> waiter;
```

Comment:
This type is `std::optional<std::variant<AsyncWaiter, SyncWaiter>>` which is ok. Alternatively one could instead use the `std::monostate` and have `std::variant<std::monostate, AsyncWaiter, SyncWaiter>`, which reads as either _nothing_ or a _AsyncWaiter_ or a _SyncWaiter_. However, I don't know if this interacts well with the serialization framework.

Reply:
This might also reduce the code clutter else where when unpacking the types.

Reply:
Instead of using `std::monostate` you can use any other empty type. For example on could create a `NoWaiter` class. The variant is default constructible if the first type is default constructible.

Reply:
Good point, I like it, much better readable.